stl/inc/format

```diff
@@ -389,26 +555,256 @@ public:
+    }
+
+protected:
+    _Basic_format_specs<_CharT> _Specs;
+};
+
+// Actually fetch the value of an argument associated with a dynamic
+// width or precision specifier. This will be called with either
+// _Width_checker or _Precision_checker as "_Handler".
+template <class _Handler, class _FormatArg>
+constexpr unsigned int _Get_dynamic_specs(_FormatArg _Arg) {
+    unsigned long long _Val = visit_format_arg(_Handler(), _Arg);
+    if (_Val > (_STD numeric_limits<unsigned int>::max)()) {
+        throw format_error("Number is too big.");
+    }
+    return static_cast<unsigned int>(_Val);
+}
+
+template <class _Context>
+constexpr basic_format_arg<_Context> _Get_arg(_Context _Ctx, size_t _Arg_id) {
+    // note: while this is parameterized on the _Arg_id type in libfmt we don't
+    // need to do that in std::format because it's only called with either an integer
+    // id or a named id (which we do not support in std::format)
+    auto _Arg = _Ctx.arg(_Arg_id);
+    if (!_Arg) {
+        throw format_error("Argument not found.");
+    }
+    return _Arg;
+}
+
+// Checks that the type and value of an argument associated with a dynamic
+// width specifier is valid.
+class _Width_checker {
+public:
+    template <class _Ty>
+    constexpr unsigned long long operator()(_Ty _Value) {
+        if constexpr (is_integral_v<_Ty>) {
+            if constexpr (is_signed_v<_Ty>) {
+                if (_Value < 0) {
+                    throw format_error("Negative width.");
+                }
+            }
+            return static_cast<unsigned long long>(_Value);
+        }
+        throw format_error("Width is not an integer.");
+    }
+};
+
+// Checks that the type and value of an argument associated with a dynamic
+// precision specifier is valid.
+class _Precision_checker {
+public:
+    template <class _Ty>
+    constexpr unsigned long long operator()(_Ty _Value) {
+        if constexpr (is_integral_v<_Ty>) {
+            if constexpr (is_signed_v<_Ty>) {
+                if (_Value < 0) {
+                    throw format_error("Nagative precision.");
+                }
+            }
+            return static_cast<unsigned long long>(_Value);
+        }
+        throw format_error("Precision is not an integer.");
+    }
+};
+
+// Parses standard format specs into a _Basic_format_specs using _Specs_setter, and,
+// in addition handles dynamic width and precision. This is seperate from _Specs setter
+// because it needs to know about the current basic_format_parse_context and basic_format_context
+// in order to fetch the width from the arguments.
+template <class _ParseContext, class _Context>
+class _Specs_handler : public _Specs_setter<typename _Context::char_type> {
+public:
+    using _CharT = typename _Context::char_type;
+
+    constexpr _Specs_handler(_Basic_format_specs<_CharT>& _Specs, _ParseContext& _Parse_ctx, _Context& _Ctx)
+        : _Specs_setter<_CharT>(_Specs), _Parse_ctx(_Parse_ctx), _Ctx(_Ctx) {}

```

Comment:
I conventionally almost always shadow members like when when I'm just going to be using them in a member initializer list.

I guess it could cause problems if the constructor ever becomes more than just a member initializer list